### PR TITLE
Dualstack CNS and azure-ipam changes

### PR DESCRIPTION
**Pull request by rjdenney**
**Reason for Change**:

This PR is to address changes in CNS and the azure-ipam to allow for a dualstack implementation. 

**Issue Fixed**:

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests

**Notes**:
Currently still testing the azure-ipam/cilium side of changes

### CODE REVIEW

File: azure-ipam/ipconfig/ipconfig.go [hunk 1]
```diff
@@ -39,18 +39,24 @@ func CreateIPConfigReq(args *cniSkel.CmdArgs) (cns.IPConfigRequest, error) {
...
+	podIPNets = make([]netip.Prefix, len(resp.PodIpInfo))
+
+	for i := range resp.PodIpInfo {
+		podCIDR := fmt.Sprintf(
+			"%s/%d",
+			resp.PodIpInfo[i].PodIPConfig.IPAddress,
+			resp.PodIpInfo[i].NetworkContainerPrimaryIPConfig.IPSubnet.PrefixLength,
+		)
+		podIPNet, err := netip.ParsePrefix(podCIDR)
+		podIPNets[i] = podIPNet
```
Comment (ashvindeodhar) [inline azure-ipam/ipconfig/ipconfig.go:53]: there is no point in assigning this if err is non-nil. This should come after line 56

File: cilium/cilium_helm_values.yaml [hunk 2]
```diff
@@ -22,10 +22,12 @@ endpointRoutes:
 
 extraArgs:
   - --local-router-ipv4=169.254.23.0
+  - --local-router-ipv6=fe80::1
```
Comment (ashvindeodhar) [inline cilium/cilium_helm_values.yaml:25]: I am just curious on where did you get this from?

File: cni/network/invoker_cns.go [hunk 3]
```diff
@@ -53,6 +53,7 @@ func NewCNSInvoker(podName, namespace string, cnsClient cnsclient, executionMode
 	}
 }
 
+// ryand IPv6 not currently supported
```
Comment (ashvindeodhar) [inline cni/network/invoker_cns.go:56]: not sure I understand. Which scenario is this code path covering? Don't we support ipv6 in CNS now with overlay?

File: cni/network/invoker_cns.go [hunk 4]
```diff
@@ -85,13 +86,13 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
...
 
 	info := IPv4ResultInfo{
-		podIPAddress:       response.PodIpInfo.PodIPConfig.IPAddress,
-		ncSubnetPrefix:     response.PodIpInfo.NetworkContainerPrimaryIPConfig.IPSubnet.PrefixLength,
-		ncPrimaryIP:        response.PodIpInfo.NetworkContainerPrimaryIPConfig.IPSubnet.IPAddress,
-		ncGatewayIPAddress: response.PodIpInfo.NetworkContainerPrimaryIPConfig.GatewayIPAddress,
-		hostSubnet:         response.PodIpInfo.HostPrimaryIPInfo.Subnet,
-		hostPrimaryIP:      response.PodIpInfo.HostPrimaryIPInfo.PrimaryIP,
-		hostGateway:        response.PodIpInfo.HostPrimaryIPInfo.Gateway,
+		podIPAddress:       response.PodIpInfo[0].PodIPConfig.IPAddress,
```
Comment (ashvindeodhar) [inline cni/network/invoker_cns.go:89]: it's wrong to assume that 0th index will always hold ipv4. You need to check if the IPAddress is v4 family or not

File: cni/network/invoker_cns_test.go [hunk 5]
```diff
@@ -56,23 +56,25 @@ func TestCNSIPAMInvoker_Add(t *testing.T) {
...
 						result: &cns.IPConfigResponse{
-							PodIpInfo: cns.PodIpInfo{
-								PodIPConfig: cns.IPSubnet{
-									IPAddress:    "10.0.1.10",
-									PrefixLength: 24,
-								},
-								NetworkContainerPrimaryIPConfig: cns.IPConfiguration{
-									IPSubnet: cns.IPSubnet{
-										IPAddress:    "10.0.1.0",
+							PodIpInfo: []cns.PodIpInfo{
```
Comment (ashvindeodhar) [inline cni/network/invoker_cns_test.go:59]: need to add v6 test

File: cns/NetworkContainerContract.go [hunk 6]
```diff
@@ -441,7 +441,7 @@ type GetIPAddressStatusResponse struct {
 
 // GetPodContextResponse is used in CNS Client debug mode to get mapping of Orchestrator Context to Pod IP UUID
 type GetPodContextResponse struct {
-	PodContext map[string]string
+	PodContext map[string][]string
```
Comment (ashvindeodhar) [inline cns/NetworkContainerContract.go:444]: please add a comment explaining why there can be one orch context to many pod IPs

File: cns/client/client_test.go [hunk 7]
```diff
@@ -124,8 +124,8 @@ func getIPNetFromResponse(resp *cns.IPConfigResponse) (net.IPNet, error) {
 	)
 
 	// set result ipconfig from CNS Response Body
-	prefix := strconv.Itoa(int(resp.PodIpInfo.PodIPConfig.PrefixLength))
-	ip, ipnet, err := net.ParseCIDR(resp.PodIpInfo.PodIPConfig.IPAddress + "/" + prefix)
+	prefix := strconv.Itoa(int(resp.PodIpInfo[0].PodIPConfig.PrefixLength))
```
Comment (ashvindeodhar) [inline cns/client/client_test.go:127]: this code needs to be updated to have a loop that runs over PodIpInfo instead of hardcoding 0th index

File: cns/restserver/ipam.go [hunk 8]
```diff
@@ -86,7 +88,7 @@ func (service *HTTPRestService) requestIPConfigHandler(w http.ResponseWriter, r
 		}
 	}
 
-	reserveResp := &cns.IPConfigResponse{
+	reserveResp := cns.IPConfigResponse{
```
Comment (ashvindeodhar) [inline cns/restserver/ipam.go:91]: I remember you were experimenting with this change to see if that helps with some error you were having. Is this change needed? I wouldn't think so

File: cns/restserver/ipam.go [hunk 9]
```diff
@@ -102,64 +104,76 @@ var (
...
+						return nil
+					}
+				}
+				endpointInfo.IfnameToIPMap[ipconfigRequest.Ifname].IPv4 = append(endpointInfo.IfnameToIPMap[ipconfigRequest.Ifname].IPv4, ipconfig)
 			}
-			endpointInfo.IfnameToIPMap[ipconfigRequest.Ifname].IPv4 = append(endpointInfo.IfnameToIPMap[ipconfigRequest.Ifname].IPv4, ipconfig)
-		}
 
-		service.EndpointState[ipconfigRequest.InfraContainerID] = endpointInfo
+			service.EndpointState[ipconfigRequest.InfraContainerID] = endpointInfo
```
Comment (ashvindeodhar) [inline cns/restserver/ipam.go:147]: there seems be extra white space on line 147

File: cns/restserver/ipam.go [hunk 10]
```diff
@@ -478,31 +499,33 @@ func (service *HTTPRestService) MarkExistingIPsAsPendingRelease(pendingIPIDs []s
...
 	defer service.RUnlock()
 
-	ipID := service.PodIPIDByPodInterfaceKey[podInfo.Key()]
-	if ipID != "" {
-		if ipState, isExist := service.PodIPConfigState[ipID]; isExist {
-			err := service.populateIPConfigInfoUntransacted(ipState, &podIpInfo)
-			return podIpInfo, isExist, err
-		}
+	for _, ipID := range service.PodIPIDByPodInterfaceKey[podInfo.Key()] {
+		if ipID != "" {
```
Comment (ashvindeodhar) [inline cns/restserver/ipam.go:512]: remove extra white spaces

File: cns/restserver/ipam.go [hunk 11]
```diff
@@ -478,31 +499,33 @@ func (service *HTTPRestService) MarkExistingIPsAsPendingRelease(pendingIPIDs []s
...
 	}
 
 	return podIpInfo, isExist, nil
 }
 
-func (service *HTTPRestService) AssignDesiredIPConfig(podInfo cns.PodInfo, desiredIPAddress string) (cns.PodIpInfo, error) {
-	var podIpInfo cns.PodIpInfo
+func (service *HTTPRestService) AssignDesiredIPConfig(podInfo cns.PodInfo, desiredIPAddress string) ([]cns.PodIpInfo, error) {
+	var podIpInfo []cns.PodIpInfo
+	podIpInfo = make([]cns.PodIpInfo, 2)
```
Comment (ashvindeodhar) [inline cns/restserver/ipam.go:528]: need to remove hardcoded value of 2

File: cns/restserver/ipam.go [hunk 12]
```diff
@@ -526,42 +549,73 @@ func (service *HTTPRestService) AssignDesiredIPConfig(podInfo cns.PodInfo, desir
 			default:
 				return podIpInfo, errors.Errorf("[AllocateDesiredIPConfig] Desired IP is not available %+v", ipConfig)
 			}
-			err := service.populateIPConfigInfoUntransacted(ipConfig, &podIpInfo)
+			err := service.populateIPConfigInfoUntransacted(ipConfig, &podIpInfo[0])
```
Comment (ashvindeodhar) [inline cns/restserver/ipam.go:552]: why is this assigning to 0th index only?

File: cns/restserver/ipam.go [hunk 13]
```diff
@@ -526,42 +549,73 @@ func (service *HTTPRestService) AssignDesiredIPConfig(podInfo cns.PodInfo, desir
...
 	return podIpInfo, fmt.Errorf("Requested IP not found in pool")
 }
 
-func (service *HTTPRestService) AssignAnyAvailableIPConfig(podInfo cns.PodInfo) (cns.PodIpInfo, error) {
+// ryand setting to default dualstack mode for testing while waiting to get pod labels - will change
+func (service *HTTPRestService) AssignAnyAvailableIPConfig(podInfo cns.PodInfo) ([]cns.PodIpInfo, error) {
 	service.Lock()
 	defer service.Unlock()
+	var podIpInfo []cns.PodIpInfo
+	podIpInfo = make([]cns.PodIpInfo, 2)
```
Comment (ashvindeodhar) [inline cns/restserver/ipam.go:564]: same comment as above

File: cns/restserver/ipam.go [hunk 14]
```diff
@@ -526,42 +549,73 @@ func (service *HTTPRestService) AssignDesiredIPConfig(podInfo cns.PodInfo, desir
...
-
-			return podIPInfo, nil
+			break
+		}
+	}
+	JsonData, err := json.Marshal(&podIpInfo[0])
+	if err != nil {
+		return []cns.PodIpInfo{}, err
+	}
+	logger.Warnf(string(JsonData))
```
Comment (ashvindeodhar) [inline cns/restserver/ipam.go:586]: why is this printed as warning?

File: cns/restserver/ipam.go [hunk 15]
```diff
@@ -526,42 +549,73 @@ func (service *HTTPRestService) AssignDesiredIPConfig(podInfo cns.PodInfo, desir
...
-			err := service.populateIPConfigInfoUntransacted(ipConfig, &podIpInfo)
+			err := service.populateIPConfigInfoUntransacted(ipConfig, &podIpInfo[0])
 			return podIpInfo, err
 		}
 	}
 	return podIpInfo, fmt.Errorf("Requested IP not found in pool")
 }
 
-func (service *HTTPRestService) AssignAnyAvailableIPConfig(podInfo cns.PodInfo) (cns.PodIpInfo, error) {
+// ryand setting to default dualstack mode for testing while waiting to get pod labels - will change
```
Comment (ashvindeodhar) [inline cns/restserver/ipam.go:559]: this logic needs to change so that it gives out one IP from each available NC. There is no need to be concerned about v4 / v6

File: cns/restserver/restserver.go [hunk 16]
```diff
@@ -58,13 +58,13 @@ type HTTPRestService struct {
...
 	homeAzMonitor            *HomeAzMonitor
 	networkContainer         *networkcontainers.NetworkContainers
-	PodIPIDByPodInterfaceKey map[string]string                    // PodInterfaceId is key and value is Pod IP (SecondaryIP) uuid.
-	PodIPConfigState         map[string]cns.IPConfigurationStatus // Secondary IP ID(uuid) is key
-	IPAMPoolMonitor          cns.IPAMPoolMonitor
-	routingTable             *routes.RoutingTable
-	store                    store.KeyValueStore
-	state                    *httpRestServiceState
-	podsPendingIPAssignment  *bounded.TimedSet
+	PodIPIDByPodInterfaceKey map[string][]string // PodInterfaceId is key and value is Pod IP (SecondaryIP) uuid.
```
Comment (ashvindeodhar) [inline cns/restserver/restserver.go:61]: update the comment to reflect the change

File: cns/restserver/restserver.go [hunk 17]
```diff
@@ -106,7 +106,7 @@ type GetHTTPServiceDataResponse struct {
 
 // HTTPRestServiceData represents in-memory CNS data in the debug API paths.
 type HTTPRestServiceData struct {
-	PodIPIDByPodInterfaceKey map[string]string                    // PodInterfaceId is key and value is Pod IP uuid.
+	PodIPIDByPodInterfaceKey map[string][]string                  // PodInterfaceId is key and value is Pod IP uuid.
```
Comment (ashvindeodhar) [inline cns/restserver/restserver.go:109]: same comment as above

File: cns/restserver/ipam_test.go [hunk 18]
```diff
@@ -67,26 +67,28 @@ func NewPodState(ipaddress string, prefixLength uint8, id, ncid string, state ty
...
+	}
 
 	// retrieve podinfo from orchestrator context
 	podInfo, err := cns.UnmarshalPodInfo(req.OrchestratorContext)
 	if err != nil {
 		return cns.IPConfigurationStatus{}, errors.Wrap(err, "failed to unmarshal pod info")
 	}
 
-	ipID := svc.PodIPIDByPodInterfaceKey[podInfo.Key()]
+	ipID := svc.PodIPIDByPodInterfaceKey[podInfo.Key()][0]
```
Comment (ashvindeodhar) [inline cns/restserver/ipam_test.go:91]: I don;t see a single ipv6 test in this PR.

**Review by ashvindeodhar — CHANGES_REQUESTED** [review]

File: azure-ipam/ipam.go [hunk 19]
```diff
@@ -85,14 +85,23 @@ func (p *IPAMPlugin) CmdAdd(args *cniSkel.CmdArgs) error {
 		p.logger.Error("Failed to interpret CNS IPConfigResponse", zap.Error(err), zap.Any("response", resp))
 		return cniTypes.NewError(ErrProcessIPConfigResponse, err.Error(), "failed to interpret CNS IPConfigResponse")
 	}
-	p.logger.Debug("Parsed pod IP", zap.String("podIPNet", podIPNet.String()))
+	for _, IPNet := range *podIPNet {
+		p.logger.Debug("Parsed pod IP", zap.String("podIPNet", IPNet.String()))
+	}
 
+	// ryand hardcoding until we can use labels
```
[the remark this answers is not in the document]
Comment (rjdenney) [reply azure-ipam/ipam.go:92]: I think this was a comment added back when we're discussing node labels and is now out of date. I'll remove it.

File: cilium/cilium_helm_values.yaml (hunk 2 above)
Comment (rjdenney) [reply cilium/cilium_helm_values.yaml:25]: So for the local router ipv4 we use the link-local address which is that line above it. When I was talking with Will he mentioned that is something I would have to add and also needs to be added on the AKS side. 

I looked up what is used for an ipv6 link-local address and found here https://blogs.infoblox.com/ipv6-coe/fe80-1-is-a-perfectly-valid-ipv6-default-gateway-address/ that it needs to be an address with a header of fe80.

File: cni/network/invoker_cns.go (hunk 4 above)
Comment (rjdenney) [reply cni/network/invoker_cns.go:89]: This may be able to be changed back to normal or will have to be something different if we end up using the new API. I originally was in the mindset that the first ip would be ipv4 since that is how I originally had it set up as and if there was only one IP it owuld default to the first index anyways. Will still need to be changed though.

File: cni/network/invoker_cns.go (hunk 3 above)
Comment (rjdenney) [reply cni/network/invoker_cns.go:56]: I am going to remove this since it is the CNI and relates more to what is being done on Windows. Should have been removed earlier.

**Review by ashvindeodhar — CHANGES_REQUESTED** [review]

File: azure-ipam/ipam_test.go [hunk 20]
```diff
@@ -31,10 +31,12 @@ func (c *MockCNSClient) RequestIPAddress(ctx context.Context, ipconfig cns.IPCon
 	case "failProcessCNSResp":
 		result := &cns.IPConfigResponse{
 			PodIpInfo: cns.PodIpInfo{
-				PodIPConfig: cns.IPSubnet{
-					IPAddress:    "10.0.1.10.2", // invalid ip address
-					PrefixLength: 24,
-				},
+				{
+					PodIPConfig: cns.IPSubnet{
+						IPAddress:    "10.0.1.10.2", // invalid ip address
```
[the remark this answers is not in the document]
Comment (rjdenney) [reply azure-ipam/ipam_test.go:36]: There is an additional .2 at the end which would make it 40 bytes. A valid IP would stop at 10.0.1.10

File: azure-ipam/ipam_test.go [hunk 21]
```diff
@@ -31,10 +31,12 @@ func (c *MockCNSClient) RequestIPAddress(ctx context.Context, ipconfig cns.IPCon
 	case "failProcessCNSResp":
 		result := &cns.IPConfigResponse{
 			PodIpInfo: cns.PodIpInfo{
```
[the remark this answers is not in the document]
Comment (rjdenney) [reply azure-ipam/ipam_test.go:33]: This test is using the old version of RequestIPAddress which expects the old contract without the slice

File: cns/restserver/ipam_test.go [hunk 22]
```diff
@@ -105,38 +112,60 @@ func NewPodStateWithOrchestratorContext(ipaddress, id, ncid string, state types.
...
-	for _, ipconfig := range ipconfigs {
-		secIpConfig := cns.SecondaryIPConfig{
-			IPAddress: ipconfig.IPAddress,
-			NCVersion: -1,
+func UpdatePodIpConfigState(t *testing.T, svc *HTTPRestService, ipconfigs map[string]cns.IPConfigurationStatus, ncIDs []string) error {
+	// Create each NC 
+	for _, NCID := range ncIDs {
+		secondaryIPConfigs := make(map[string]cns.SecondaryIPConfig)
+		for _, ipconfig := range ipconfigs {
+			if ipconfig.NCID == NCID {
```
[the remark this answers is not in the document]
Comment (rjdenney) [reply cns/restserver/ipam_test.go:120]: So this was a part of the updates that I had to make to the ipam_test.go suite. There may be a way to do it but what this part of the function does is fill up the NCs with the ipconfigs that are passed in. 

The outer loop create an NC for each NCID that I pass in. The inner loop then goes through the ipconfigs passed in and searches for all of the ipconfigs that are assigned to that NC. By checking if the ipconfig equals the NCID we are on we can keep IPs assigned to only the NC we want them to be on.

What I eventually did later down the suite was make all of the tests able to have variable inputs for IPs, NCs, and prefixes. Each test should be able to use those to create whatever pods needed to run the test and can work for single or multiple NC cases.

File: cns/restserver/internalapi_test.go [hunk 23]
```diff
@@ -581,28 +581,28 @@ func validateNCStateAfterReconcile(t *testing.T, ncRequest *cns.CreateNetworkCon
...
+			ipConfigstate := svc.PodIPConfigState[ipID]
+			if ipConfigstate.GetState() != types.Assigned {
+				t.Fatalf("IpAddress %s is not marked as assigned to Pod: %+v, ipState: %+v", ipaddress, podInfo, ipConfigstate)
+			}
 
-		// Validate if IPAddress matches
-		if ipConfigstate.IPAddress != ipaddress {
-			t.Fatalf("IpAddress %s is not same, for Pod: %+v, actual ipState: %+v", ipaddress, podInfo, ipConfigstate)
-		}
+			// Validate if IPAddress matches
```
[the remark this answers is not in the document]
Comment (rjdenney) [reply cns/restserver/internalapi_test.go:591]: This section got indented because I had to add an extra for loop for ranging over svc.PodIPIDByPodInterfaceKey[podInfo.Key()]  since the map's value is now a slice of IPs instead of just an IP.
